private/pkg/git/tree.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+type tree struct {
+	hash    Hash
+	entries []TreeEntry
+}
+
+func (t *tree) Hash() Hash {
+	return t.hash
+}
+
+func (t *tree) Entries() []TreeEntry {
+	return t.entries
+}
+
+func (t *tree) Traverse(objectReader ObjectReader, names ...string) (TreeEntry, error) {
+	return traverse(objectReader, t, names...)
+}
+
+func parseTree(hash Hash, data []byte) (*tree, error) {
+	t := &tree{
+		hash: hash,
+	}
+	for len(data) > 0 {
+		i := bytes.Index(data, []byte{0})
+		if i == -1 {
+			return nil, errors.New("malformed tree")
+		}
+		length := i + 1 + hashLength
+		entry, err := parseTreeEntry(data[:length])
```

Comment:
could use some comments as to the math here - overall, we could use with a bunch of inline documentation for how this all works, for ourselves to refer to in two years heh